server.go

```diff
@@ -336,6 +341,54 @@ func (s *Server) EnforceRetentionPolicies() {
 	}
 }
 
+// StartShardGroupsPreCreate launches shard group pre-create to avoid write bottlenecks.
+func (s *Server) StartShardGroupsPreCreate(checkInterval time.Duration) error {
+	if checkInterval == 0 {
+		return fmt.Errorf("shard group pre-create check interval must be non-zero")
+	}
+	sgpcDone := make(chan struct{}, 0)
+	s.sgpcDone = sgpcDone
+	go func() {
+		for {
+			select {
+			case <-sgpcDone:
+				return
+			case <-time.After(checkInterval):
+				s.ShardGroupPreCreate(checkInterval)
+			}
+		}
+	}()
+	return nil
+}
+
+// ShardGroupPreCreate ensures that future shard groups and shards are created and ready for writing
+// is removed from the server.
+func (s *Server) ShardGroupPreCreate(checkInterval time.Duration) {
+	log.Println("shard group pre-create check commencing")
+
+	// For safety, we double the check interval to ensure we have enough time to create all shard groups
+	// before they are needed, but as close to needed as possible.
+	// This is a complete punt on optimization
+	cutoff := time.Now().Add(checkInterval * 2).UTC()
+
+	// Check all shard groups.
+	// See if they have a "future" shard group ready to write to
+	// If not, create the next shard group, as well as each shard for the shardGroup
+	for _, db := range s.databases {
+		for _, rp := range db.policies {
+			for _, g := range rp.shardGroups {
+				// Check to see if it is going to end before our interval
+				if g.EndTime.Before(cutoff) {
+					log.Printf("pre-creating shard group for %d, retention policy %s, database %s", g.ID, rp.Name, db.name)
+					if err := s.PreCreateShardGroupIfNotExists(db.name, rp.Name, g.ID); err != nil {
```

Comment:
You will then need to send these messages outside of this lock. Check out retention policy enforcement.